pkg/controller/seed-controller-manager/cluster-credentials-controller/controller.go

```diff
@@ -0,0 +1,197 @@
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	ControllerName = "kkp-cluster-credentials-controller"
+)
+
+type Reconciler struct {
+	ctrlruntimeclient.Client
+
+	workerName string
+	recorder   record.EventRecorder
+	log        *zap.SugaredLogger
+	versions   kubermatic.Versions
+}
+
+// Add creates a new cluster-credentials controller.
+func Add(
+	mgr manager.Manager,
+	numWorkers int,
+	workerName string,
+	log *zap.SugaredLogger,
+	versions kubermatic.Versions,
+) error {
+	reconciler := &Reconciler{
+		Client: mgr.GetClient(),
+
+		workerName: workerName,
+		recorder:   mgr.GetEventRecorderFor(ControllerName),
+		log:        log,
+		versions:   versions,
+	}
+
+	c, err := controller.New(ControllerName, mgr, controller.Options{
+		Reconciler:              reconciler,
+		MaxConcurrentReconciles: numWorkers,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create controller: %w", err)
+	}
+
+	if err := c.Watch(&source.Kind{Type: &kubermaticv1.Cluster{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return fmt.Errorf("failed to create watch: %w", err)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.With("cluster", request.Name)
+	log.Debug("Reconciling")
+
+	cluster := &kubermaticv1.Cluster{}
+	if err := r.Get(ctx, request.NamespacedName, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// do not migrate a cluster in deletion
+	if cluster.DeletionTimestamp != nil {
+		return reconcile.Result{}, nil
+	}
+
+	// Add a wrapping here so we can emit an event on error
+	result, err := kubermaticv1helper.ClusterReconcileWrapper(
+		ctx,
+		r.Client,
+		r.workerName,
+		cluster,
+		r.versions,
+		kubermaticv1.ClusterConditionNone,
+		func() (*reconcile.Result, error) {
+			return r.reconcile(ctx, log, cluster)
+		},
+	)
+	if err != nil {
+		log.Errorw("Failed to reconcile cluster", zap.Error(err))
+		r.recorder.Event(cluster, corev1.EventTypeWarning, "ReconcilingError", err.Error())
+	}
+	if result == nil {
+		result = &reconcile.Result{}
+	}
+	return *result, err
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, cluster *kubermaticv1.Cluster) (*reconcile.Result, error) {
+	oldCluster := cluster.DeepCopy()
+
+	// patch in the cleanup finalizer first (the pkg/clusterdeletion takes care of cleaning up)
+	kuberneteshelper.AddFinalizer(cluster, kubermaticv1.CredentialsSecretsCleanupFinalizer)
+
+	if !equality.Semantic.DeepEqual(oldCluster.Finalizers, cluster.Finalizers) {
+		if err := r.Patch(ctx, cluster, ctrlruntimeclient.MergeFrom(oldCluster)); err != nil {
+			return nil, fmt.Errorf("failed to add finalizer: %w", err)
+		}
+
+		return &reconcile.Result{Requeue: true}, nil
```

Comment:
why are the requeue's after patches needed?

Reply:
They are not strictly necessary, but since we will cause a retrigger anyway because of the updates, it makes sense to just skip any further effort and wait for the next reconciliation. Ultimately this prevents a conflict later on (that is glossed over by our automatic retrying).